core/drivers/clk/fixed_clk.c

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <stdint.h>
+
+struct fixed_clock_data {
+	unsigned long rate;
+};
+
+static unsigned long fixed_clk_get_rate(struct clk *clk,
+					unsigned long parent_rate __unused)
+{
+	struct fixed_clock_data *d = clk->priv;
+
+	return d->rate;
+}
+
+static const struct clk_ops fixed_clk_clk_ops = {
+	.get_rate = fixed_clk_get_rate,
+};
+
+static TEE_Result fixed_clock_setup(const void *fdt, int offs)
+{
+	const uint32_t *freq = NULL;
+	const char *name = NULL;
+	struct clk *clk = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct fixed_clock_data *fcd = NULL;
+
+	name = fdt_get_name(fdt, offs, NULL);
+	if (!name)
+		name = "fixed-clock";
+
+	clk = clk_alloc(name, &fixed_clk_clk_ops, NULL, 0);
+
+	fcd = calloc(1, sizeof(struct fixed_clock_data));
+	if (!clk || !fcd)
+		return TEE_ERROR_OUT_OF_MEMORY;
```

Comment:
must free the allocated resource before returning. Or could panic()!

Ditto for the other error cases below (lines 48, 55 and 57).

Reply:
Added free/clk_free